Add XHR and Same-origin Policy (#230)
browserengineering/book: book/security.md

all: if someone stole your `token` cookie, they could copy it into
their browser, and the server would think they are you.

Our browser must prevent one servers from seeing *another server's*

> cookie values.[^11][^12][^13][^14] But attackers might be able to get

Reviewer: Optional: rename these footnotes if you're up for it.

Cross-site requests
===================

Imagine that your browser is logged in to your bank, and than an
attacker wants to know your (private) bank balance. Already, our
browser stores different cookies for each site. Because of this, it
won't just send the bank cookie to an attacker's site. But what if the
attacker is clever?

The easiest way for an attacker to steal your private data is to ask
for it. Of course, there's no API in the browser for a website to ask
for another website's cookies. But there _is_ an API to make requests
to another website. It's called `XMLHttpRequest`.[^weird-name]

> [^weird-name]: It's a weird name! Why is `XML` capitalized but not

Reviewer: Are you sure it can handle non-HTTP requests? I'm not sure that's correct.

Reviewer: Add: "The XML name comes from the historical origins of the API as part of the Outlook Web Access project at Microsoft (https://en.wikipedia.org/wiki/XMLHttpRequest)

Author: I did edit this footnote to include the history.

I'll remove the HTTP comment, I agree it's a stretch.

[^even-more-options]: As above, this implementation skips important
    `XMLHttpRequest` features, like setting request headers (and
    reading response headers), changing the response type, or
    triggering various events and callbacks during the request.

``` {.javascript}
XMLHttpRequest.prototype.send = function(body) {
    this.responseText = call_python("XMLHttpRequest_send",
        this.method, this.url, this.body);
}
```

In Python, implementing `send` is pretty simple. Just define and
export this `XMLHttpRequest_send` function:[^note-method]

[^note-method]: Note that the `method` argument is ignored, because
    our `request` function chooses the method on its own based on
    whether a payload is passed. This is again not what the standard
    requires, and a careful implementation would fix it.

``` {.python replace=full_url%2c/full_url%2c%20self.tab.url%2c}
class JSContext:
    def XMLHttpRequest_send(self, method, url, body):
        full_url = resolve(url, self.tab.url)
        headers, out = request(full_url, payload=body)
        return out
```

With this new JavaScript API, a web page can make HTTP requests while

> the user interacts with it, making various "live", "AJAX" websites

Reviewer: ```suggestion
the user interacts with it, making more interactive websites
```

possible! This API, and newer analogs like [`fetch`][mdn-fetch], are

> the basis for so many web page interactions that don't reload the

Reviewer: ```suggestion
the basis for many web page interactions that don't reload the
```

> page---think "liking" a post or hover previews or similar.

Reviewer: ```suggestion
page---think "liking" a post, hover previews or submitting a form without reloading the page.
```

[mdn-fetch]: https://developer.mozilla.org/en-US/docs/Web/API/fetch

Same-origin Policy
==================

However, new capabilities lead to new responsibilities. After all:
any HTTP requests triggered by `XMLHttpRequest` will include cookies,

> which means they can potentially be used to steal or abuse cookies!

Reviewer: ```suggestion
which means they can potentially be used to steal or abuse private information.
```

This is by design: when you "like" something, the corresponding
HTTP request needs your cookie so the server associates the "like" to
your account. But it also means that `XMLHttpRequest`s have access to
private data, and thus need to protect it.

Let's imagine an attacker that wants to know your username on our
guest book server. When you're logged in, the guest book includes your
username in the "Hello, so and so" header, so it's enough for the

> attacker to read the guest book web page with your cookies.

Reviewer: ```suggestion
attacker to read the guest book web page with your cookies, in order to find out your name
and violate your privacy.
```

`XMLHttpRequest` could let them do that. Say the user visits the
attacker's website[^why-visit-attackers], which then executes the
following script:

[^why-visit-attackers]: Why is the user on the attacker's site?
    Perhaps it has funny memes, or it's been hacked and is being used
    for the attack against its will, or perhaps the evil-doer paid for
    ads on sketchy websites where users have low standards for
    security anyway.

``` {.javascript}
x = new XMLHttpRequest();
x.open("GET", "http://localhost:8000/", false);
x.send();
user = x.responseText.split(" ")[2].split("<")[0];
```

The issue here is that private guest book content is being sent to the
attacker. Since the content is derived from the cookie, it leaks
private data. To prevent this, the browser must prevent the attacker's
page from reading the guest book web page content.

> The term for this is the "Same-origin policy". Basically, web pages

Reviewer: ```suggestion
This is achieved by enforcing a *same-origin policy*. Requests subject to this policy enforce that web pages
```

Reviewer: Link: https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy

> can only make `XMLHttpRequests` for web pages on the same

Reviewer: ```suggestion
 only make `XMLHttpRequests` for web pages on the same
```

Exercises
=========

*New inputs*: Add support for hidden and password input elements.
Hidden inputs shouldn't show up or take up space, while password input
elements should show ther contents as stars instead of characters.

*Cookie access*: Add support for the `HttpOnly` and `Secure`
parameters for cookies. `HttpOnly` cookies should not be available
from JavaScript's `document.cookie`, but should be sent along with
HTTP requests. `Secure` cookies should only be sent over HTTPS, but
not over HTTP, requests.  `HttpOnly` cookies provide some protection
against cross-site scripting attacks, while `Secure` cookies prevent
an attacker from reading cookies sent over unencrypted networks.

*Cookie Expiration*: Add support for cookie expiration. Cookie
expiration dates are set in the `Set-Cookie` header, and can be
overwritten if the same cookie is set again with a later date. On the
server side, save the same expiration dates in the `SESSIONS` variable
and use it to delete old sessions to save memory.

> *CORS*: Web servers can *opt in* to allowing cross-origin

Reviewer: ```suggestion
*CORS*: Web servers can [*opt in*](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) to allowing cross-origin
```

Hidden inputs shouldn't show up or take up space, while password input
elements should show ther contents as stars instead of characters.

*Cookie access*: Add support for the `HttpOnly` and `Secure`
parameters for cookies. `HttpOnly` cookies should not be available
from JavaScript's `document.cookie`, but should be sent along with
HTTP requests. `Secure` cookies should only be sent over HTTPS, but
not over HTTP, requests.  `HttpOnly` cookies provide some protection
against cross-site scripting attacks, while `Secure` cookies prevent
an attacker from reading cookies sent over unencrypted networks.

*Cookie Expiration*: Add support for cookie expiration. Cookie
expiration dates are set in the `Set-Cookie` header, and can be
overwritten if the same cookie is set again with a later date. On the
server side, save the same expiration dates in the `SESSIONS` variable
and use it to delete old sessions to save memory.

*CORS*: Web servers can [*opt in*](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) to allowing cross-origin
`XMLHttpRequest`s using the `Access-Control-Allow-Origin` header. The
way it works is that on cross-origin HTTP requests, the web browser

> sends an `Origin` header with the origin of the requesting site. By

Reviewer: Also explain that such requests to the server will contain a cookie from the origin of the get request URL.

default, the browser then throws away the response to prevent private
data from leaking. But if the server sends the
`Access-Control-Allow-Origin` header, and its value is either the
requesting origin or the special `*` value, the browser instead makes
the output available to the script. All requests made by your browser
will be what the CORS standard calls "simple requests".

*Referer*: When your browser visits a web page, or when it loads a CSS

> or JavaScript file, it sends a `Referer` header[^24] containing the

Reviewer: ```suggestion
or JavaScript file, it sends a `Referer` header[^referer] containing the
```

Author: Thanks. Yep, the initial footnotes were auto-exported from Org-mode, and it's good to use descriptive names as we edit.

`same-origin` (only do so if navigating to another page on the same
origin). Implement those two values for `Referer-Policy`.

> [^24]: Yep, spelled that way.

Reviewer: ```suggestion
[^referer]: Yep, spelled that way.
```
